Add: remove_trailing_repeat_consonants()

konbraphat51 opened the pull request:
### What does this changes
Add removement method of repeating consonants by dictionary-based method.


### How this fixes it
Add pythainlp.utils.remove_repeat_consonant(text, dictionary)

Find words in the dictionary that has repeating consonant at the last, and find the match with the end of the sentence. If there is none, make the repetition to one; if there is, make the repetition to that one.

Fixes #860

### Your checklist for this pull request
🚨Please review the [guidelines for contributing](https://github.com/PyThaiNLP/pythainlp/blob/dev/CONTRIBUTING.md) to this repository.

- [x] Passed code styles and structures
- [x] Passed code linting checks and unit test

konbraphat51 committed:
Merge pull request #3 from PyThaiNLP/dev

update

konbraphat51 committed:
documentation

konbraphat51 committed:
Add: implemation

tested

konbraphat51 committed:
Add: test code

konbraphat51 committed:
Add: remove_repeat_consonants()

function to remove consonants
เริ่ดดดดดดดด -> เริ่ด

implementation + test code written.
Test passed

pep8speaks commented:
Hello @konbraphat51! Thanks for updating this PR. We checked the lines you've touched for [PEP 8](https://www.python.org/dev/peps/pep-0008) issues, and found:



* In the file [`pythainlp/util/remove_trailing_repeat_consonants.py`](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py):

> [Line 51:69](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py#L51): [W291](https://duckduckgo.com/?q=pep8%20W291) trailing whitespace
> [Line 126:9](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py#L126): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 128:9](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py#L128): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 249:13](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py#L249): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator
> [Line 249:50](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py#L249): [E203](https://duckduckgo.com/?q=pep8%20E203) whitespace before ':'
> [Line 251:13](https://github.com/PyThaiNLP/pythainlp/blob/3315cb026fc2e2b8b4518c3ae3a9b04c50738b10/pythainlp/util/remove_trailing_repeat_consonants.py#L251): [W503](https://duckduckgo.com/?q=pep8%20W503) line break before binary operator





##### Comment last updated at 2023-11-13 07:50:03 UTC

konbraphat51 committed:
Merge branch 'dev' of https://github.com/konbraphat51/pythainlp into dev

konbraphat51 committed:
Fix: push miss

this is the right commit

konbraphat51 committed:
Fix: divide the exceeding length code

konbraphat51 committed:
Refac: remove last white space

konbraphat51 committed:
Fix: restrict only to consonants

konbraphat51 committed:
Refac: Remove unused import

konbraphat51 commented:
Hmm... I cannot find more PEP8 error,,,
I will try more next morning. Good night!

konbraphat51 committed:
Refac: Use enumerate

pointed out by codacy

konbraphat51 committed:
Fix: add the function in init

konbraphat51 committed:
Refac: use black

konbraphat51 committed:
Refac: repeatedly used black

"1 file left unchanged." shown

konbraphat51 commented:
Lint issue of normalize.py solved

konbraphat51 committed:
Refac: resolve nested if

pointed out by codeclimate

konbraphat51 committed:
Fix test case

konbraphat51 committed:
Refac: seperate function

Cognitive complexity pointed out by CodeClimate

Black used

konbraphat51 committed:
Refac: reduce line length

pointed out by Lint
black used

konbraphat51 committed:
Refac: seperate 2 functions

konbraphat51 committed:
Refac: use black

vscode autopep8 and black has been conflicting. So autopep8 cutted

konbraphat51 committed:
Refac: seperate match finding method

cognitive complexity pointed out by CodeClimate.
Black used.

konbraphat51 committed:
Improve: save consonants repeaters for improve speed

TODO resolved, black used, test passed

konbraphat51 committed:
Refac: make repeater checking function

Code complexity pointed out by CodeClimate, black used

konbraphat51 commented:
Okey, Lint problem and CodeClimate problem are basically solved, and the test for this passed.
Review please!

It seems that the code length of a single file is set to 250 lines, but should I separate this method from normalize.py?

bact commented:
> It seems that the code length of a single file is set to 250 lines, but should I separate this method from normalize.py?

Maybe a good idea. Separate it.

Additional rational apart from the line limit is that the new function is a bit different from other existing normalization functions as well, in a way that it is no longer rule-based but need a kind of word list --> which means the normalization output will depend on what dictionary each user use each time.

From that fact, I think we shouldn't put this remove_repeat_consonant() in the list of functions to be perform inside normalize(). To keep the output of normalize() consistent all the time and independent from the provided dictionary.

User can call this new function separately.

konbraphat51 commented:
Thanks, OK

bact commented:
BTW, the proposed name of the function: `remove_repeat_consonants()` is a bit inaccurate - if compare with a similar name function: `remove_repeat_vowels()`. The latter remove repeating chars anywhere in the string but the former only remove it at the end of the string.

Naming ideas:
- `remove_trailing_repeat_consonants()` - descriptive, and long
  - consistent with `remove_repeat_vowels()`*
- `remove_trailing_dup_consonants()` - descriptive, 3 chars shorter
  - consistent with `remove_dup_spaces()`**
- `rstrip_consonants()` - sounds bold
  - borrowed from `rstrip()`
- others?

*, **  gotcha. the current code has naming inconsistency, need to clean up.

konbraphat51 committed:
Refac: seperate function

konbraphat51 committed:
Improve: Rename method

suggested by https://github.com/PyThaiNLP/pythainlp/pull/862#issuecomment-1805830606

konbraphat51 committed:
Refac: make names more clear

konbraphat51 committed:
Refac: reflect method name change

konbraphat51 committed:
Fix: argument inconsistence

konbraphat51 committed:
Merge: seperate + rename

seperate + rename

konbraphat51 committed:
Refac: revert to the first place

because this PR inplemention seperated from normalize.py

konbraphat51 committed:
Refac: use black

line-length=79

konbraphat51 committed:
Refac: reduce col length

used black line-length=79

konbraphat51 commented:
* Files seperated
* `remove_trailing_repeat_consonants()` sounds good to me, so I changed the name to that.
* I think I clarified names in the script, but if there are still name inconsistencies, please specify them.
* I revert normalize.py to what it was, so the Lint is pointing it out now
* Not sure why codeacy saying "pythainlp.util.remove_trailing_repeat_consonants.remove_trailing_repeat_consonants' imported but unused."  It is used in test code. Any idea?

bact commented on pythainlp/util/normalize.py:
```diff
@@ -299,4 +299,4 @@ def maiyamok(sent: Union[str, List[str]]) -> List[str]:
         _list_word.append(text)
         i += 1
-    return _list_word
+    return _list_word
```
Please add newline to the end of file.

Refer to this POSIX standard
https://pubs.opengroup.org/onlinepubs/9699919799/basedefs/V1_chap03.html#tag_03_206

See some examples for illustration at https://stackoverflow.com/questions/729692/why-should-text-files-end-with-a-newline

konbraphat51 replied:
Oh, my mistake.
Fixed!

bact commented:
> I think I clarified names in the script, but if there are still name inconsistencies, please specify them.

Sorry. I was meant to say that the existing `remove_repeat_vowels()` and  `remove_dup_spaces()` have naming inconsistency issue. Not from the new function you're now contributing.

bact commented on pythainlp/util/__init__.py:
```diff
@@ -103,6 +103,9 @@
     reorder_vowels,
 )
+from pythainlp.util.remove_trailing_repeat_consonants import (
+    remove_trailing_repeat_consonants,
```
I think you also need to add `remove_trailing_repeat_consonants` name to `__all__` (at the very top of the file) as well.

Names in `__all__` will be exported when `from <module> import *` is used.

This is why codeacy/pyflakes said the function is unused -- as it is imported but has never been used/exported.

konbraphat51 replied:
Oh, I didn’t know that feature…
Fixed!

bact replied:
Thank you.
(I never know it either until I start to maintain this lib lol.)

bact commented on pythainlp/util/remove_trailing_repeat_consonants.py:
```diff
@@ -0,0 +1,256 @@
+# See the License for the specific language governing permissions and
+# limitations under the License.
+"""
+Removement of repeated consonants
```
Suggestion:
Removement of repeated consonants at the end of words

konbraphat51 replied:
Fixed!

bact commented on pythainlp/util/remove_trailing_repeat_consonants.py:
```diff
@@ -0,0 +1,256 @@
+# contains all words that has repeating consonants at the end
+# for each consonant
+# when dictionary updated, this should be updated too
+# key: consonan
```
key: consonant

konbraphat51 replied:
Fixed!

bact commented on pythainlp/util/remove_trailing_repeat_consonants.py:
```diff
@@ -0,0 +1,256 @@
+from pythainlp import thai_consonants as consonants
+from typing import Tuple, List
+
+# used by remove_repeat_consonants()
```
used by remove_trailing_repeat_consonants()

konbraphat51 replied:
Forgot updating..
Fixed!

bact commented:
Neat contribution, thank you. I think that's all. I can approve after few suggested fixes above.

konbraphat51 committed:
Refac: add last new line

konbraphat51 committed:
Update commentation

Update responding to method rename

konbraphat51 commented:
Sorry, mistakely “start review” button pushed please wait🥲

konbraphat51 committed:
Refac: clearify commentation

konbraphat51 commented:
Accidentally pushed “start review” button, sorry, please wait for a while🥲

konbraphat51 committed:
Refac: fix typi

konbraphat51 committed:
Add: __all__

konbraphat51 commented:
Ok, all done.
Review please :)

bact committed:
Sort export names in __all__

bact commented:
I have a last minute thought about which name we should go with,
between:
- `remove_trailing_repeat_xxx` vs
- `remove_repeat_trailing_xxx`

(try to think what it will like in the context of code completion, when ppl type `remove_` .. which char they)

But I will merge this first.
We can later thinking more about consistent names (together with that `remove_dup_xxx`) and change this again when appropriate.

Thank you for your contribution.

(the test on Windows is now running for 58 minutes and counting, this shouldn't be the norm :( )
